libsolidity/codegen/ArrayUtils.cpp

```diff
@@ -823,6 +823,94 @@ void ArrayUtils::incrementDynamicArraySize(ArrayType const& _type) const
 		})", {"ref"});
 }
 
+void ArrayUtils::popStorageArrayElement(ArrayType const& _type) const
+{
+	solAssert(_type.location() == DataLocation::Storage, "");
+	solAssert(_type.isDynamicallySized(), "");
+	if (!_type.isByteArray() && _type.baseType()->storageBytes() < 32)
+		solAssert(_type.baseType()->isValueType(), "Invalid storage size for non-value type.");
+
+	if (_type.isByteArray())
+	{
+		m_context.appendInlineAssembly(R"({
+			let slot_value := sload(ref)
+			switch and(slot_value, 1)
+			case 0 {
+				// short byte array
+				let length := and(div(slot_value, 2), 0x3f)
+				if iszero(length) { invalid() }
+
+				// Zero-out the suffix of the byte array by masking it.
+				// Do not zero-out the least significant byte, but mask the
+				// higher bits of the length.
+				// (((1<<(8 * (32 - length))) - 1) << 8) + 192
+				let mask := add(mul(0x100, sub(exp(0x100, sub(32, length)), 1)), 0xc0)
+				slot_value := and(not(mask), slot_value)
```

Comment:
Perhaps the following would be easier to understand:

```
let mask := mul(0x100, sub(exp(0x100, sub(32, length)), 1))
length := sub(length, 1)
slot_value := or(and(not(mask), slot_value), mul(length, 2))
```

I.e. just mask away the least significant byte and then or-in the adjusted length.